Framework/Algorithms/src/CalcCountRate.cpp

```diff
@@ -0,0 +1,653 @@
+      // initialize thread's histogram buffer
+      nThreads = PARALLEL_NUMBER_OF_THREADS;
+      Buff.resize(nThreads);
+      for (int i = 0; i < nThreads; i++) {
+        Buff[i].assign(m_numLogSteps, 0);
+      }
+    }
+
+#pragma omp for
+    for (int64_t i = 0; i < nHist; ++i) {
+      auto nThread = PARALLEL_THREAD_NUMBER;
+      PARALLEL_START_INTERUPT_REGION
+
+      // Get a const event list reference. eventInputWS->dataY() doesn't work.
+      const DataObjects::EventList &el = InputWorkspace->getSpectrum(i);
+      el.generateCountsHistogramPulseTime(dTRangeMin, dTRangeMax, Buff[nThread],
+                                          m_XRangeMin, m_XRangeMax);
+      if (this->buildVisWS()) {
+        this->hisogramEvents(el, pVisWS_locks.get());
+      }
+
+      // Report progress
+      prog.report(name());
+      PARALLEL_END_INTERUPT_REGION
+    }
+    PARALLEL_CHECK_INTERUPT_REGION
+// calculate final sums
+#pragma omp for
+    for (int64_t j = 0; j < m_numLogSteps; j++) {
+
+      for (int i = 0; i < nThreads; i++) {
+        countRate[j] += Buff[i][j];
+      }
+      // normalize if requested
+      if (!countNormalization.empty()) {
+        countRate[j] /= countNormalization[j];
+      }
+    }
+    if (!countNormalization.empty() && this->buildVisWS()) {
+#pragma omp for
+      for (int64_t j = 0; j < int64_t(m_visNorm.size()); j++) {
+        this->normalizeVisWs(j);
+      }
+    }
+  }
+  // recover histogram timing, but start assuming run starts at
+  std::vector<Kernel::DateAndTime> times(m_numLogSteps);
+
+  double dt = (dTRangeMax - dTRangeMin) / static_cast<double>(m_numLogSteps);
+  auto t0 = m_TRangeMin.totalNanoseconds();
+  for (auto i = 0; i < m_numLogSteps; i++) {
+    times[i] =
+        Kernel::DateAndTime(t0 + static_cast<int64_t>((0.5 + double(i)) * dt));
+  }
+  // store calculated values within the target log.
+  targLog->replaceValues(times, countRate);
+}
+/** hisogram event list into visualization workspace
+* @param el       :: event list to rebin into visualization workspace
+* @param spectraLocks :: pointer to the array of mutexes to lock modifyed
+*                        visualization workspace spectra for a thread
+*/
+void CalcCountRate::hisogramEvents(const DataObjects::EventList &el,
+                                   std::mutex *spectraLocks) {
+
+  if (el.empty())
+    return;
+
+  auto events = el.getEvents();
+  for (const DataObjects::TofEvent &ev : events) {
+    double pulsetime = static_cast<double>(ev.pulseTime().totalNanoseconds());
+    double tof = ev.tof();
+    if (pulsetime < m_visT0 || pulsetime >= m_visTmax)
+      continue;
+    if (tof < m_XRangeMin || tof >= m_XRangeMax)
+      continue;
+
+    size_t n_spec = static_cast<size_t>((pulsetime - m_visT0) / m_visDT);
+    size_t n_bin = static_cast<size_t>((tof - m_XRangeMin) / m_visDX);
+    (spectraLocks + n_spec)->lock();
+    auto &Y = m_visWs->mutableY(n_spec);
+    Y[n_bin]++;
+    (spectraLocks + n_spec)->unlock();
```

Comment:
I share you concern but after careful consideration think it will never happens. Workspace is allocated within the program and should be well verified.